Fix Sentry breadcrumbs collection during initialization

Gudahtt opened the pull request:
## Explanation

The refactor of the Sentry state in #20491 accidentally broke our opt- in detection in the `beforeBreadcrumbs` hook for errors in the background process. It would have broken our error report opt-in detection as well, except that that function was broken in a way that made it work correctly but inefficiently (it was loading the persisted state each time to check, ignoring the application state).

The opt-in detection has been updated to look for both types of application state (during and after initialization). The function was refactored to be used for the `beforeBreadcrumbs` hook as well.

## Manual Testing Steps

This was discovered by @danjm when investigating the e2e test failure in https://github.com/MetaMask/metamask-extension/pull/20517. This can be tested by following a similar approach to the failing e2e test: modify the persisted state in a way that breaks a migration, and ensure the breadcrumbs are collected.

The steps I took are:
* Comment out this line: https://github.com/MetaMask/metamask-extension/blob/88212a7c82ce20f37b5949246d3751a0be605cf4/app/scripts/lib/setupSentry.js#L140
* Create a development build with a valid `SENTRY_DSN_DEV` set in `.metamaskrc`
* Run this code snippet in the background console after installing the extension:
```
chrome.storage.local.get(({ data, meta }) => chrome.storage.local.set({ data: { ...data, MetaMetricsController: { ...data.MetaMetricsController, participateInMetaMetrics: true }, OnboardingController: { ...data.OnboardingController, completedOnboarding: true }, NftController: false }, meta: {...meta, version: 20} }, () => { window.location.reload() }))
```
  * This snippet sets the version back far enough that there are plenty of console logs/breadcrumbs to collect
  * `onboardingComplete` is set to `true` and `participateInMetaMetrics` is set to `true` so that the request goes through and breadcrumbs aren't suppressed
* See that there is a resulting Sentry error report in the network tab, and breadcrumbs are included in the payload
* Try again with `onboardingComplete` set to `false`, and you should see the same error report but without the breadcrumbs
* Try again with `participateInMetaMetrics` set to `false` and there should be no error report
* After initialization, try running `window.stateHooks.throwTestError()` in the UI (checking the network tab to see whether an error is sent, and whether it has breadcrumbs, which should depend on the two properties referenced earlier)
* After initialization, try running `window.stateHooks.throwTestBackgroundError()` in the UI (checking the network tab to see whether an error is sent, and whether it has breadcrumbs, which should depend on the two properties referenced earlier)

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

Gudahtt committed:
Fix Sentry MetaMetrics detection

The refactor of the Sentry state in #20491 accidentally broke our opt-
in detection. The opt-in detection has been updated to look for both
types of application state (during and after initialization).

danjm commented on app/scripts/lib/setupSentry.js:
```diff
@@ -321,17 +359,12 @@ function hideUrlIfNotInternal(url) {
-      if (
-        Object.values(appState).length &&
-        (!appState?.store?.metamask?.participateInMetaMetrics ||
-          !appState?.store?.metamask?.completedOnboarding ||
```
We also need to account for the `!appState?.store?.metamask?.completedOnboarding` in the refactor. Perhaps by adding another new function like `getMetaMetricsEnabledFromAppState`

Gudahtt replied:
What is the purpose of checking that property?

Gudahtt replied:
Good catch, fixed in a8b1dafdd3

I left it inlined here for now since it's the only place we check it

Gudahtt committed:
Continue suppressing breadcrumbs during onboarding

danjm commented on app/scripts/lib/setupSentry.js:
```diff
@@ -321,17 +359,15 @@ function hideUrlIfNotInternal(url) {
+    const appState = getState();
+    if (
+      !getMetaMetricsEnabledFromAppState(appState) ||
+      !appState?.store?.metamask?.completedOnboarding ||
```
Do we need to handle the case where `appState.store` is undefined, but`appState.persistedState` is defined?

Gudahtt replied:
Yep, just found this while testing too. Fixed here I think: aabb66be35 (still testing)

danjm commented on app/scripts/lib/setupSentry.js:
```diff
@@ -127,6 +127,45 @@ export const SENTRY_UI_STATE = {
+  } else if (appState.state) {
+    // UI
+    if (appState.state.metamask) {
+      return Boolean(appState.state.metamask?.participateInMetaMetrics);
```
Will `appState.state.metamask` be always undefined? Because of https://github.com/MetaMask/metamask-extension/pull/20428/files#diff-215d2cf6fb64ee62b9f484e7bbdd33246739e39b854e3cc8c2245b026352c12a

Gudahtt replied:
`metamask` is only present because that name is from the UI Redux store, `metamask` is the name of the background state Redux slice. Previously we used to imitate the UI state in the background on the line you just linked, but that was undone in that PR.

The background state has controllers as top-level keys, so `metamask` should never be present.

Gudahtt committed:
Fix how onboarding status is retrieved

The check for whether the user had completed onboarding assumed that
the application state was post-initialization UI state. It has been
updated to handle background state and pre-initialization state as
well.

Gudahtt commented:
OK, test instructions updated, I think it works now. I was thrown off at first because I wasn't seeing any breadcrumbs, but that was because I was only moving the state back to migration 91, which is far enough to produce console warnings but not far enough to guarantee they're picked up by Sentry in time to be included.

danjm reviewed (dismissed):
LGTM!

Gudahtt commented on app/scripts/lib/setupSentry.js:
```diff
@@ -127,6 +127,71 @@ export const SENTRY_UI_STATE = {
+  // during initialization after loading persisted state
+  if (appState.persistedState) {
+    return Boolean(
+      appState.persistedState.data.OnboardingController?.completedOnboarding,
```
Missing `?`, we don't know that `data` is an object
```suggestion
      appState.persistedState.data?.OnboardingController?.completedOnboarding,
```

Gudahtt commented on app/scripts/lib/setupSentry.js:
```diff
@@ -127,6 +127,71 @@ export const SENTRY_UI_STATE = {
+  } else if (appState.state) {
+    // UI
+    if (appState.state.metamask) {
+      return Boolean(appState.state.metamask?.completedOnboarding);
```
Unnecessary `?`, the condition ensures that this is truthy
```suggestion
      return Boolean(appState.state.metamask.completedOnboarding);
```

Gudahtt commented on app/scripts/lib/setupSentry.js:
```diff
@@ -127,6 +127,71 @@ export const SENTRY_UI_STATE = {
+/**
+ * Returns whether onboarding has completed, given the application state.
+ *
+ * @param {unknown} appState - Application state
```
We do actually know that it's an object at this point (otherwise `appState.persistedState` would crash in that first condition)
```suggestion
 * @param {Record<string, unknown>} appState - Application state
```

Gudahtt commented on app/scripts/lib/setupSentry.js:
```diff
@@ -127,6 +127,71 @@ export const SENTRY_UI_STATE = {
+  } else if (appState.state) {
+    // UI
+    if (appState.state.metamask) {
+      return Boolean(appState.state.metamask?.participateInMetaMetrics);
```
Unnecessary `?`, we checked `appState.state.metamask` already in the condition
```suggestion
      return Boolean(appState.state.metamask.participateInMetaMetrics);
```

Gudahtt committed:
Remove unnecessary optional chain operators

Gudahtt committed:
Add missing optional chain operator

Gudahtt committed:
Fix JSDoc description parameter type

danjm reviewed (approved):
LGTM

NidhiKJha reviewed (approved):
LGTM! PR description is written very nicely 🤩
